### PR TITLE
Conditions support for picture-elements inner elements

### PR DESCRIPTION
**Pull request by yosilevy**
This was requested a while back by a bunch of people. This can allow to take picture-elements in to the next level (I think this is needed in many other cards as well.
For example it can help create an interactive multi device remote. Now people just show/hide whole cards and that's bad UX...

Here's just a simple demo of just one control:
https://www.dropbox.com/s/h6t2hlrbj07qs94/Capture.mp4?dl=0

As this is my first card update please let me know how to modify docs? And demo card? What else needs to be done?

### CODE REVIEW

**Comment by iantrich** [comment]
Cool 👍

File: src/panels/lovelace/cards/hui-picture-elements-card.ts [hunk 1]
```diff
@@ -94,6 +111,18 @@ class HuiPictureElementsCard extends LitElement implements LovelaceCard {
     `;
   }
 
+  private _evalConditions(elementConfig: LovelaceElementConfig): boolean {
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-picture-elements-card.ts]: We shouldn't built this into the picture elements card, instead it should be a conditional element, like the conditional card.

Comment (iantrich) [reply src/panels/lovelace/cards/hui-picture-elements-card.ts]: I'm pretty sure you can already do it with the existing conditional card; only caveat is that you would use a "custom" element for the card. e.g. 

type: custom:hui-icon-element

**Comment by yosilevy** [comment]
But how do you do it for elements and not for a complete card? The idea is to do it on elements. This is much more powerful and conditions should be a basic capability also in glance cards etc...

File: src/panels/lovelace/cards/hui-picture-elements-card.ts (hunk 1 above)
Comment (yosilevy) [reply src/panels/lovelace/cards/hui-picture-elements-card.ts]: I think I know what @balloob means. Will try.
@iantrich - the idea is to do this at the element level within picture elements. I'll check if hui-conditional-card would work. 
Anyway something that is not part of the card will be missed by many users and it's a shame IMHO.

Comment (yosilevy) [reply src/panels/lovelace/cards/hui-picture-elements-card.ts]: @iantrich so I tried this (see below) and off course it fails since picture-elements inner elements are not cards. Am I missing something?
Later once I have time I will go with what Paulus suggested to wrap it in a conditional element that hosts other elements.

```
image: 'https://home.consiv.com/local/garden.jpg'
type: picture-elements
elements:
  - type: custom:hui-conditional-card
    conditions:
      - entity: input_boolean.bool1
        state: on
    card:
    - type: state-label
      entity: sensor.solar_angle
```

Comment (yosilevy) [reply src/panels/lovelace/cards/hui-picture-elements-card.ts]: Done. Refactored to be a new conditional element that allows sub elements.

**Comment by iantrich** [comment]
You've duplicated the conditional-card logic into element, this is a bad idea for maintenance. Instead, create a hui-conditional component that allows for cards/rows/elements to be used with it. hui-conditional-card is then just a wrapper around hui-conditional component; same being true for row and element. We then have shared logic to maintain and gain the ability to add conditionals in rows and elements.

**Comment by yosilevy** [comment]
Code is not clean. Just for review. Please see comments as to why some things were done.

File: src/panels/lovelace/cards/hui-conditional-card.ts [hunk 2]
```diff
@@ -1,79 +1,22 @@
...
-    }
-    // This will hide the complete card so it won't get styled by parent
-    this.style.setProperty("display", visible ? "" : "none");
+    // todo: artifically using the above config but it's still important to distinguish between card config and the internal hui-conditional implementation config
+    super.setConditionalConfig(config as Config);
   }
 
   public getCardSize() {
-    return computeCardSize(this._card!);
+    return super.getCardSize();
```
Comment (iantrich) [inline src/panels/lovelace/cards/hui-conditional-card.ts]: card size would only be relevant to the conditional card, not rows or elements, so not necessary to be in the base class

**Comment by balloob** [comment]
I don't like how this is structured, I find it very hard to follow.

File: src/panels/lovelace/elements/hui-conditional-element.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    super.setConditionalConfig(config);
+  }
+
+  protected getCustomStyle(): string {
+    return `
+        :host {
+          position: absolute;
+          left: 0px;
+          top: 0px;
+          width: 100%;
```
Comment (balloob) [inline src/panels/lovelace/elements/hui-conditional-element.ts]: This seems like a hack. What is going on.

File: src/panels/lovelace/elements/hui-conditional-element.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        .element {
+          position: absolute;
+          transform: translate(-50%, -50%);
+        }
+    `;
+  }
+
+  // this is here and not in hui-conditional because otherwise there is a circular reference (hui-conditional needs this
+  // class to create sub elements but it is also needed to create the conditional element itself created by this class)
+  protected _createHuiElement(
```
Comment (balloob) [inline src/panels/lovelace/elements/hui-conditional-element.ts]: Better to extract in a helper method? Otherwise it will get out of sync.

File: src/panels/lovelace/components/hui-conditional.ts [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+import { EntityRowConfig } from "../entity-rows/types";
+import { createCardElement } from "../common/create-card-element";
+// import { createRowElement } from "../common/create-row-element";
+
+export interface ConditionalConfig {
+  conditions: Condition[];
+  card: LovelaceCardConfig;
+  elements: LovelaceElementConfig[];
+  rows: EntityRowConfig[];
+  // todo: add row (can't find any createhuirow or LovelaceRowConfig...)
```
Comment (balloob) [inline src/panels/lovelace/components/hui-conditional.ts]: MVP, don't do rows.

File: src/panels/lovelace/components/hui-conditional.ts [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+      !config.conditions.every((c) => c.entity && (c.state || c.state_not)) ||
+      (!config.card && (!config.elements || !Array.isArray(config.elements)))
+    ) {
+      throw new Error("Error in conditional configuration.");
+    }
+
+    this._conditionalConfig = config;
+
+    // set style (overridable)
+    // LitElement static styles can't work since we want to allow deriving classes to modify it
```
Comment (balloob) [inline src/panels/lovelace/components/hui-conditional.ts]: Not true, static styles can just add super.styles to their CSS array. Make sure the return set is set to `CSSResultArray`

File: src/panels/lovelace/components/hui-conditional.ts [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+    this.removeItems();
+
+    this.createItems();
+
+    if (this._hass) {
+      this.hass = this._hass;
+    }
+  }
+
+  protected getCustomStyle(): string {
```
Comment (balloob) [inline src/panels/lovelace/components/hui-conditional.ts]: Remove.

File: src/panels/lovelace/components/hui-conditional.ts (hunk 6 above)
Comment (balloob) [reply src/panels/lovelace/components/hui-conditional.ts]: Actually, why would we even need this to begin with ? The deriving classes need to handle CSS, this class should not be aware. 

Also, a conditional card really should not have ANY CSS. It's only showing something else.

File: src/panels/lovelace/components/hui-conditional.ts [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+  conditions: Condition[];
+  card: LovelaceCardConfig;
+  elements: LovelaceElementConfig[];
+  rows: EntityRowConfig[];
+  // todo: add row (can't find any createhuirow or LovelaceRowConfig...)
+  // rows: Lovelace;
+  // use it with createRowElement
+}
+
+export class HuiConditional extends LitElement {
```
Comment (balloob) [inline src/panels/lovelace/components/hui-conditional.ts]: Why is this a `LitElement` ? The original conditional card is derived from an HTMLElement because there is no lit logic needed/used.

File: src/panels/lovelace/components/hui-conditional.ts [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+        }
+      );
+    }
+    // todo: add rows, etc'....
+  }
+
+  private handleItems(): boolean {
+    const conditionsMet =
+      this._conditionalConfig &&
+      this._conditionalConfig.conditions.every((c) => {
```
Comment (balloob) [inline src/panels/lovelace/components/hui-conditional.ts]: This should be extracted into a helper method, just so it can easily be tested etc.

File: src/panels/lovelace/components/hui-conditional.ts [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+    }
+
+    return conditionsMet as boolean;
+  }
+
+  private appendItems() {
+    if (this._card) {
+      this._card.hass = this._hass;
+      if (!this._card!.parentNode) {
+        this.shadowRoot!.appendChild(this._card);
```
Comment (balloob) [inline src/panels/lovelace/components/hui-conditional.ts]: The original conditional card was not using a shadow root, neither should this element.

File: src/panels/lovelace/components/hui-conditional.ts [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+      this._elements.map((el: LovelaceElement) => {
+        if (el.parentNode) {
+          this.shadowRoot!.removeChild(el);
+        }
+      });
+    }
+  }
+
+  private hasNoItems() {
+    return !this._card && (!this._elements || this._elements.length === 0);
```
Comment (balloob) [inline src/panels/lovelace/components/hui-conditional.ts]: This is so dirty. If you are trying to create a generic conditional card, why is there knowledge in here about cards or elements? That's not generic at all.

File: src/panels/lovelace/components/hui-conditional.ts [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+  }
+}
+
+declare global {
+  interface HTMLElementTagNameMap {
+    "hui-conditional": HuiConditional;
+  }
+}
+
+customElements.define("hui-conditional", HuiConditional);
```
Comment (balloob) [inline src/panels/lovelace/components/hui-conditional.ts]: Why define this, where is it used? It's not. Mark the class as abstract.

**Comment by balloob** [comment]
Looked at it, still too complicated.

**Comment by yosilevy** [comment]
As discussed I will be reverting this to my original implementation + extracting some helper functions.

**Comment by sauloonze** [comment]
Hello guys. Thanks for the contribution here! I am bringing a condition issue here. 
I am not using hass and getting a "TypeError: Cannot convert undefined or null to object"

**Comment by yosilevy** [comment]
You should not use this code. It is being refactored and will be re-introduced later on
